reframe/core/schedulers/sge.py

```diff
@@ -0,0 +1,225 @@
+            if opt.startswith('-'):
+                rem_opts.append(opt)
+            elif opt.startswith('#'):
+                verb_opts.append(opt)
+            else:
+                select_opt += ':' + opt
+
+        return [self._format_option(select_opt),
+                *(self._format_option(opt) for opt in rem_opts),
+                *verb_opts]
+
+    def _format_option(self, option):
+        return self._prefix + ' ' + option
+
+    def make_job(self, *args, **kwargs):
+        return _SgeJob(*args, **kwargs)
+
+    def emit_preamble(self, job):
+        preamble = [
+            self._format_option('-N "%s"' % job.name),
+            self._format_option('-o %s' % job.stdout),
+            self._format_option('-e %s' % job.stderr),
+            self._format_option('-wd %s' % job.workdir),
+        ]
+
+        if job.time_limit is not None:
+            h, m, s = seconds_to_hms(job.time_limit)
+            preamble.append(
+                self._format_option('-l h_rt=%d:%d:%d' % (h, m, s)))
+
+        preamble += self._emit_lselect_option(job)
+
+        return preamble
+
+    def allnodes(self):
+        raise NotImplementedError('sge backend does not support node listing')
+
+    def filternodes(self, job, nodes):
+        raise NotImplementedError('sge backend does not support '
+                                  'node filtering')
+
+    def submit(self, job):
+        # `-o` and `-e` options are only recognized in command line by the PBS,
+        # SGE, and Slurm wrappers.
+        cmd = f'qsub -o {job.stdout} -e {job.stderr} {job.script_filename}'
+        completed = _run_strict(cmd, timeout=self._submit_timeout)
+        jobid_match = re.search(r'^Your job (?P<jobid>\S+)', completed.stdout)
+        if not jobid_match:
+            raise JobSchedulerError('could not retrieve the job id '
+                                    'of the submitted job')
+
+        job._jobid = jobid_match.group('jobid')
+        job._submit_time = time.time()
+
+    def wait(self, job):
+        intervals = itertools.cycle([1, 2, 3])
+        while not self.finished(job):
+            self.poll(job)
+            time.sleep(next(intervals))
+
+    def cancel(self, job):
+        time_from_submit = time.time() - job.submit_time
+        if time_from_submit < SGE_CANCEL_DELAY:
+            time.sleep(SGE_CANCEL_DELAY - time_from_submit)
+
+        _run_strict(f'qdel {job.jobid}', timeout=self._submit_timeout)
+        job._cancelled = True
+
+    def finished(self, job):
+        if job.exception:
+            raise job.exception
+
+        return job.completed
+
+    def poll(self, *jobs):
+        if jobs:
+            # Filter out non-jobs
+            jobs = [job for job in jobs if job is not None]
+
+        if not jobs:
+            return
+
+        user = osext.osuser()
+        completed = osext.run_command(
+            f'qstat -xml -u {user}'
+        )
+
+        if completed.returncode != 0:
```

Comment:
I am not familiar with SGE so I am just checking here. Does squeue return information for old jobs indefinitely? If not, what is the return code in case (a) all the job ids are old (b) one of the job ids is old? I am looking at PBS and it returns (a) status code 153 and (b) status code 0, but only includes recent IDs in the output.

Reply:
> Does squeue return information for old jobs indefinitely?

SGE's `qstat` doesn't show old jobs at all, as far as I can see.  [`qacct`](https://linux.die.net/man/1/qacct) shows old jobs but not active ones (so it can't be used as replacement for `qstat`), and it can be _slow_, depending on whether how often (if ever) sysadmins rotate flat files.

> If not, what is the return code in case (a) all the job ids are old (b) one of the job ids is old? I am looking at PBS and it returns (a) status code 153 and (b) status code 0, but only includes recent IDs in the output.

I'm not passing the list of jobs to `qstat`.

Reply:
> I'm not passing the list of jobs to `qstat`.

To elaborate on this, the reason why I'm simply calling [`qstat`](https://arc.liv.ac.uk/SGE/htmlman/htmlman1/qstat.html) with all jobs for the current user is because SGE's `qstat` is very weird, if you call it with `qstat -f -j <comma-separated list of jobs>` you get _less_ information about the job itself, for example you don't get its status.

Reply:
Ok, I understand now, thanks! Then it is fine I think to ask for all the jobs of the user 👍